docs/src/index.md

```diff
@@ -6,4 +6,369 @@ DocTestSetup = :(using ChainRulesCore, ChainRules)
+And they learn how to combine simple rules, via the chain rule, to differentiate complicated functions.
+ChainRules.jl is a programmatic repository of that knowledge, with the generalizations to higher dimensions.
+
+Autodiff (AD) tools roughly work by reducing a problem down to simple parts that they know the rules for,
+and then combining those rules.
+Knowing rules for more complicated functions speeds up the autodiff process as it doesn't have to break things down as much.
+
+** ChainRules is an AD independent collection of rules to use in a differentiation system **
+
+### `rrule` and `frule`
+
+!!! Terminology "`rrule` and `frule`"
+
+    `rrule` and `frule` are ChainRules.jl specific terms.
+    And there exact functioning is kind of ChainRule specific,
+    though other tools may do similar.
+    The core notion is sometimes called
+    _Custom AD primitives_, _custom adjoints_, _custom sensitivities_.
+
+The rules are encoded as `rrules` and `frules`,
+for use in reverse-mode and forward-mode differentiation respectively.
+
+The `rrule` for some function `foo`, which takes the positional argument `args` and keyword argument `kwargs`, is written:
+```julia
+function rrule(::typeof(foo), args; kwargs...)
+    ...
+    return y, pullback
+end
+```
+where `y` must be equal to `foo(args; kwargs...)`,
+and _pullback_ is a function to propagate the derivative information backwards at that point (more later).
+Often but not always it is calculated directly.
+The exception is that we can calculate it indirectly to make
+the `pullback` faster. (more on _pullback_ later)
+
+Similarly, the `frule` is written:
+```julia
+function frule(::typeof(foo), args; kwargs...)
+    ...
+    return y, pushforward
+end
+```
+again `y=foo(args, kwargs...)`,
+and _pushforward_ is a function to propagate the derivative information forwards at that point (more later).
+
+Almost always the _pushforward_/_pullback_ will be declared locally with-in the `ffrule`/`rrule`, and will be a _closure_ over some of the other arguments.
+
+### The propagators: pushforward and pullback
+
+!!! Terminology "Pushforward and Pullback"
+
+    _Pushforward_ and _Pullback_ are fancy words that the autodiff community recently stole from Differential Geometry.
+    The are broadly in agreement with the use of these terms in differential geometry. But any geometer will tell you these are the super-boring flat cases. Some will also frown at you.
+    Other terms that may be used include for _pullback_ the **backpropagator**, and by analogy for _pushforward_ the **forwardpropagator**, thus these are the _propagators_.
+    These are also good names because effectively they propagate wibbles and wobbles through them, via the chainrule.
+    (the term **backpropagator** may originate with ["Lambda The Ultimate Backpropagator"](http://www-bcl.cs.may.ie/~barak/papers/toplas-reverse.pdf) by Bearlmutter and Siskind, 2008)
+
+
+#### Core Important Idea:
```

Comment:
Are these ideas consistent with what `pushforward` and `pullback` do? I'm not familiar with ChainRules and its internals, but I anticipated `pushforward` and `pullback` to do the following: Consider a computation `x -> u -> f(u) = v -> y`. Then `pushforward` for `f` turns `du/dx` into `dv/dx`, whereas `pullback` turns `dy/dv` into `dy/du`. So `pushforward` pushes a "sensitivity with respect to the input through the function", whereas `pullback` pulls a "sensitivity with respect to the output back through the function". Perhaps that's what the below convey, not sure... maybe I'm just rambling.

Reply:
Yeah, I think the below is accurate for the pushforward but misleading for the pullback. The pullback doesn’t take an output wobble and produce an input wiggle (that would be left-multiplying by the _inverse_ of the Jacobian); it takes an output sensitivity (“how much does the loss function wobble when you wiggle the output”) and produces an input sensitivity (“how much does the loss function wobble when you wiggle the input”). This corresponds to left-multiplying by the _adjoint_ of the Jacobian—an important distinction!

If the output _is_ the scalar loss and you call the pullback on the scalar 1, then it will produce the _gradient_ of the input (also a vector in the cotangent space, aka a wobble-wiggle ratio).